### PR TITLE
Added Sky Mall to Dwarven Overlay

### PR DESCRIPTION
**Pull request by hannibal002**
image

### CODE REVIEW

File: src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            }
+
+            return prefix + name
+        }
+
+        fun daySuffix(day: Int) = when (day) {
+            1 -> "st"
+            2 -> "nd"
+            3 -> "rd"
+            else -> "th"
```
Comment (NopoTheGamer) [inline src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt:115]: isnt this wrong for day 21 and above?

File: src/main/kotlin/io/github/moulberry/notenoughupdates/guifeatures/SkyMallDisplay.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            SkyMallVariant.values().find { it.chatMessage == message }
+        }.also {
+            if (it != null) {
+                currentVariant = it
+                println("found ${it.displayName}")
+                update()
+            } else {
+                println("nothing found!")
+            }
+        }
```
Comment (romangraef) [inline src/main/kotlin/io/github/moulberry/notenoughupdates/guifeatures/SkyMallDisplay.kt:54]:
```suggestion
        val message = matcher.group(1) ?: return
        currentVariant = SkyMallVariant.values().find { it.chatMessage == message }
        update()
```

File: src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+class SkyBlockTime constructor(
+    var year: Int = 1,
+    var month: Int = 1,
+    var day: Int = 1,
+    var hour: Int = 0,
+    var minute: Int = 0,
+    var second: Int = 0,
+) {
+
+    fun toRealTime(): Long {
```
Comment (romangraef) [inline src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt:31]: Consider using a stronger type like `Instant` here

File: src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+package io.github.moulberry.notenoughupdates.util
+
+class SkyBlockTime constructor(
+    var year: Int = 1,
+    var month: Int = 1,
+    var day: Int = 1,
+    var hour: Int = 0,
+    var minute: Int = 0,
+    var second: Int = 0,
+) {
```
Comment (romangraef) [inline src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt:29]: Consider using a data class:
```suggestion
data class SkyBlockTime(
    val year: Int = 1,
    val month: Int = 1,
    val day: Int = 1,
    val hour: Int = 0,
    val minute: Int = 0,
    val second: Int = 0,
) {
```
This auto generates a `toString()` method as well as a `copy()` method (in place of your `duplicate()` method. In addition making the fields `val` still allows you to change the fields by using `.copy(hour = 0, minutes = 0  /* etc */)` with all the benefits of immutability.
(Data classes can have either `val` or `var` fields, if you prefer having a mutable data class)

**Review by romangraef — CHANGES_REQUESTED** [review]

File: src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt (hunk 4 above)
Comment (hannibal002) [reply src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt:29]: I wanted to make the primary constructor private. That's the intention behind this. Additionally, the fields have to be mutable for the logic I want to use them for. Similar to Calendar

Comment (hannibal002) [reply src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt]: implemented.

File: src/main/kotlin/io/github/moulberry/notenoughupdates/guifeatures/SkyMallDisplay.kt (hunk 2 above)
Comment (hannibal002) [reply src/main/kotlin/io/github/moulberry/notenoughupdates/guifeatures/SkyMallDisplay.kt:54]: yeah i dont need the debugs anymore ur right

Comment (hannibal002) [reply src/main/kotlin/io/github/moulberry/notenoughupdates/guifeatures/SkyMallDisplay.kt]: Implemented.

File: src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt (hunk 1 above)
Comment (hannibal002) [reply src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt:104]: Implemented.

**Comment by hannibal002** [comment]
This PR is ready (except `Instant` but im waiting on Nea on this)

**Review by romangraef — DISMISSED** [review]

File: src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt [hunk 5]
```diff
@@ -82,7 +82,7 @@ data class SkyBlockTime(
                 else -> "Undefined!"
             }
 
-            val name = when (month / 4 + 1) {
```
Comment (romangraef) [inline src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt]: I'm probably gonna revert this commit cause it is in my unit test changes already.

Comment (hannibal002) [reply src/main/kotlin/io/github/moulberry/notenoughupdates/util/SkyBlockTime.kt]: 🆗

**Review by romangraef — APPROVED** [review]

**Review by NopoTheGamer — CHANGES_REQUESTED** [review]
11st
12nd
13rd

**Comment by NopoTheGamer** [comment]
image
Is it meant to be coloured and capitalised like this? it looks strange

**Comment by hannibal002** [comment]
> 11st 12nd 13rd

At this point I guess it's better when you fix it instead 😂

**Review by romangraef — DISMISSED** [review]

**Review by NopoTheGamer — APPROVED** [review]

**Review by romangraef — APPROVED** [review]

**Review by romangraef — APPROVED** [review]

**Pull request merged**